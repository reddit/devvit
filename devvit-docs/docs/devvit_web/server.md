# @devvit/server

The `@devvit/server` package provides core server-side functionality for Devvit Web apps. It allows you to create HTTP servers, manage request context, and configure your Devvit app with a traditional client/server architecture.

## Overview

The `@devvit/server` package works seamlessly with Devvit Web, providing:

- **HTTP Server Creation**: Create and configure HTTP servers for your Devvit Web app
- **Request Context Management**: Access request-specific context and metadata
- **App Configuration**: Define app permissions, post types, and server behavior
- **Port Management**: Handle server port configuration for different environments

## Installation

The `@devvit/server` package is included by default in Devvit Web projects. If you need to install it manually:

```bash
npm install @devvit/server
```

## API Reference

### `createServer`

Creates a new Devvit HTTP server with the same API as Node.js's `createServer` function. You must use this variant of createServer, and not the one that is available directly from Node.js.

```typescript
import { createServer } from '@devvit/server';

// Basic usage
const server = createServer((req, res) => {
  res.writeHead(200, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ message: 'Hello from Devvit!' }));
});
```

**Parameters:**

- `optionsOrRequestListener` (optional): Server options or request handler function
- `requestListener` (optional): Request handler function

**Returns:** A Node.js HTTP server instance

**Note:** The server may not actually create an HTTP server under the hood, but it will behave like one in terms of API compatibility.

### `context`

A proxy object that provides access to the current request context. This is automatically set by the server when handling requests.

```typescript
import { context } from '@devvit/server';

// In your request handler
app.get('/api/data', (req, res) => {
  // Access request context
  const postId = context.postId;
  const subredditName = context.subredditName;
  const userId = context.userId;

  // Use context data in your response
  res.json({
    postId,
    subredditName,
    userId,
  });
});
```

**Available Context Properties:**

- `postId`: The ID of the current post (if applicable)
- `subredditName`: The name of the current subreddit
- `userId`: The ID of the current user
- `userName`: The username of the current user
- `isModerator`: Whether the current user is a moderator
- `isLoggedIn`: Whether the current user is logged in

**Note:** If you try to access `context` outside of a server request, it will throw an error.

### `getServerPort`

Gets the port number that the server should run on, reading from the `WEBBIT_PORT` environment variable.

```typescript
import { createServer, getServerPort } from '@devvit/server';

const server = createServer((req, res) => {
  // Handle request
});

// Start server on the configured port
server.listen(getServerPort());
```

**Returns:** Port number (defaults to 3000 if `WEBBIT_PORT` is not set)

### `RequestContext`

Type definition for the request context object.

```typescript
import type { RequestContext } from '@devvit/server';

function handleRequest(req: Request, res: Response) {
  // RequestContext type is automatically inferred when using context
  const ctx: RequestContext = context;
}
```

## Usage Examples

### Basic Express.js Server

```typescript
import express from 'express';
import { createServer, getServerPort, context } from '@devvit/server';

const app = express();

app.get('/api/hello', (req, res) => {
  res.json({
    message: 'Hello from Devvit!',
    postId: context.postId,
    subredditName: context.subredditName,
  });
});

app.post('/api/data', express.json(), (req, res) => {
  // Handle POST request with JSON body
  res.json({ success: true });
});

const server = createServer(app);
server.listen(getServerPort());
```

### Server with Custom Middleware

```typescript
import express from 'express';
import { createServer, getServerPort, context } from '@devvit/server';

const app = express();

// Custom middleware to log request context
app.use((req, res, next) => {
  console.log('Request context:', {
    postId: context.postId,
    subredditName: context.subredditName,
    userId: context.userId,
  });
  next();
});

app.get('/api/user-info', (req, res) => {
  if (!context.isLoggedIn) {
    return res.status(401).json({ error: 'User not logged in' });
  }

  res.json({
    userId: context.userId,
    userName: context.userName,
    isModerator: context.isModerator,
  });
});

const server = createServer(app);
server.listen(getServerPort());
```

## Limitations

- **Serverless execution**: Server endpoints run in a serverless environment with limited execution time
- **No persistent connections**: WebSocket connections and long-running processes are not supported
- **Package restrictions**: Node.js packages that rely on specific native dependencies may not be available in the serverless environment
- **Request/response only**: Streaming responses and chunked transfers are not supported

## Related Documentation

- [Devvit Web Overview](./devvit_web_overview)
- [Devvit Web Quickstart](./devvit_web_quickstart)
- [@devvit/client Documentation](../client)
- [Devvit Web Templates](./devvit_web_templates)
