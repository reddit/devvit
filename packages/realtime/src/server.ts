import { RealtimeClient } from './RealtimeClient.js';

export type { RealtimeClient } from './RealtimeClient.js';
export const realtime: RealtimeClient = new RealtimeClient();
