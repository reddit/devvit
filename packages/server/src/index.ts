export { context, getContextCache, runWithContext, setContextCache } from './context.js';
export { createServer } from './create-server.js';
export { Context, setContext } from './server-context.js';
export { getServerPort } from '@devvit/shared-types/server/get-server-port.js';
