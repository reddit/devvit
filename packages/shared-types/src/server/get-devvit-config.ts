import type { Config } from '../Config.js';
import type {} from '../shared/devvit-worker-global.js';

/**
 * Gets the Config object that the runtime drops on globalThis. For internal use only.
 * @internal
 */
export function getDevvitConfig(): Config {
  if (!globalThis.devvit?.config) {
    throw new Error(
      'Devvit config is not available. Make sure to call getDevvitConfig() after the Devvit runtime has been initialized.'
    );
  }
  return globalThis.devvit.config;
}
