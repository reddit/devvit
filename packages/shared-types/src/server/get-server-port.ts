import process from 'node:process';

/**
 * Gets the port the internal server should run on. Call this when starting the server to get the
 * port to use.
 * @example createServer(myHandler).listen(getServerPort());
 * @returns The port to use for the server.
 */
export function getServerPort(): number {
  const envPort = Number(process.env.WEBBIT_PORT);
  if (isNaN(envPort)) {
    return 3000;
  }
  return envPort;
}
