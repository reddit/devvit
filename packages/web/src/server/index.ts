export * from '@devvit/cache';
export * from '@devvit/media';
export * from '@devvit/realtime/server';
export * from '@devvit/reddit';
export * from '@devvit/redis';
export * from '@devvit/scheduler';
export * from '@devvit/server';
export * from '@devvit/settings';
