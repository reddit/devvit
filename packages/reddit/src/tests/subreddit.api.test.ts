import type { Listing as ListingProto, SubredditAboutResponse_AboutData } from '@devvit/protos';
import { context } from '@devvit/server';
import { describe, expect, test, vi } from 'vitest';

import { AboutLocations, Subreddit } from '../models/Subreddit.js';
import { RedditClient } from '../RedditClient.js';
import { redditApiPlugins } from './utils/redditApiPluginsMock.js';
import { runWithTestContext } from './utils/runWithTestContext.js';
import { userActionsPlugin } from './utils/userActionsPluginMock.js';

vi.mock('../plugin.js', () => {
  return {
    getRedditApiPlugins: () => redditApiPlugins,
    getUserActionsPlugin: () => userActionsPlugin,
  };
});

vi.mock('../common.js', () => {
  return {
    assertUserScope: vi.fn(),
    RunAs: {
      APP: 0,
      USER: 1,
      UNSPECIFIED: 2,
    },
  };
});

function createTestSub(options: Partial<SubredditAboutResponse_AboutData>): Subreddit {
  return new Subreddit({
    id: 'someId',
    displayName: 'askReddit',
    subredditType: 'public',
    createdUtc: Date.parse('2022-01-03'),
    lang: 'en',
    allowedMediaInComments: [],
    userFlairRichtext: [],
    submissionType: 'any',
    ...options,
  });
}

describe('Subreddit API', () => {
  const redditAPI = new RedditClient();
  const username = 'unusual_setup';
  const subredditId = 't5_abc123';

  const mockListingWithPostsAndComments: ListingProto = {
    kind: 'Listing',
    data: {
      children: [
        {
          kind: 't1',
          data: {
            id: 'abc123',
            body: 'This is a comment',
            createdUtc: Date.parse('2022-01-01'),
            author: username,
            parentId: 't3_abc123',
            subreddit: 'pics',
            subredditId: subredditId,
            linkId: 't3_abc123',
            permalink: 'permalink://',
            allAwardings: [],
            authorFlairRichtext: [],
            awarders: [],
            treatmentTags: [],
            modPermissions: [],
            linkFlairRichtext: [],
            spoiler: false,
            modReports: [],
            userReports: [],
            gallery: [],
          },
        },
        {
          kind: 't3',
          data: {
            id: 'abc123',
            title: 'This is a post',
            createdUtc: Date.parse('2022-01-02'),
            author: username,
            subreddit: 'pics',
            subredditId: subredditId,
            url: 'url://',
            permalink: 'permalink://',
            removed: true,
            removedBy: 'spez',
            removedByCategory: 'moderator',
            allAwardings: [],
            authorFlairRichtext: [],
            awarders: [],
            treatmentTags: [],
            modPermissions: [],
            linkFlairRichtext: [],
            spoiler: false,
            modReports: [],
            userReports: [],
            gallery: [],
          },
        },
      ],
    },
  };

  describe('RedditClient:Subreddit', () => {
    test('getCommentsAndPostsByUser()', async () => {
      const spyPlugin = redditApiPlugins.Users.UserWhere;
      spyPlugin.mockImplementationOnce(async () => mockListingWithPostsAndComments);

      await runWithTestContext(async () => {
        const result = redditAPI.getCommentsAndPostsByUser({
          username,
        });

        const items = await result.all();

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            after: undefined,
            before: undefined,
            limit: 100,
            more: undefined,
            username: 'unusual_setup',
            where: 'overview',
          },
          context.metadata
        );

        expect(items).toMatchSnapshot();
      });
    });

    test('getSubredditRemovalReasons()', async () => {
      const removalReasonId = 'uuid-abc';
      const spyPlugin = redditApiPlugins.Subreddits.SubredditGetRemovalReasons;
      spyPlugin.mockImplementationOnce(async () => {
        return {
          data: {
            [removalReasonId]: {
              id: removalReasonId,
              title: 'Spam',
              message: 'This is spam!',
            },
          },
          order: [removalReasonId],
        };
      });

      await runWithTestContext(async () => {
        const result = await redditAPI.getSubredditRemovalReasons('askReddit');

        expect(spyPlugin).toHaveBeenCalledWith({ subreddit: 'askReddit' }, context.metadata);

        expect(result).toMatchSnapshot();
      });
    });

    test('addSubredditRemovalReason()', async () => {
      const spyPlugin = redditApiPlugins.Subreddits.SubredditAddRemovalReason;
      spyPlugin.mockImplementationOnce(async () => {
        return {
          id: 'uuid-abc',
        };
      });

      await runWithTestContext(async () => {
        const result = await redditAPI.addSubredditRemovalReason('askReddit', {
          title: 'Spam',
          message: 'This is spam!',
        });

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            message: 'This is spam!',
            subreddit: 'askReddit',
            title: 'Spam',
          },
          context.metadata
        );

        expect(result).toEqual('uuid-abc');
      });
    });

    test('subscribe()', async () => {
      const subreddit = new Subreddit({
        id: subredditId,
        displayName: 'askReddit',
        subredditType: 'public',
        createdUtc: Date.parse('2022-01-03'),
        lang: 'en',
        allowedMediaInComments: [],
        userFlairRichtext: [],
        submissionType: 'any',
      });
      vi.spyOn(Subreddit, 'getFromMetadata').mockImplementationOnce(async () => {
        return subreddit;
      });

      const spyPlugin = redditApiPlugins.Subreddits.Subscribe;
      spyPlugin.mockImplementationOnce(async () => {
        return {};
      });

      await runWithTestContext(async () => {
        await redditAPI.subscribeToCurrentSubreddit();

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            action: 'sub',
            actionSource: '',
            srName: 'askReddit',
            sr: '',
            skipInitialDefaults: true,
          },
          context.metadata
        );
      });
    });

    test('unsubscribe()', async () => {
      const subreddit = new Subreddit({
        id: subredditId,
        displayName: 'askReddit',
        subredditType: 'public',
        createdUtc: Date.parse('2022-01-03'),
        lang: 'en',
        allowedMediaInComments: [],
        userFlairRichtext: [],
        submissionType: 'any',
      });
      vi.spyOn(Subreddit, 'getFromMetadata').mockImplementationOnce(async () => {
        return subreddit;
      });

      const spyPlugin = redditApiPlugins.Subreddits.Subscribe;
      spyPlugin.mockImplementationOnce(async () => {
        return {};
      });

      await runWithTestContext(async () => {
        await redditAPI.unsubscribeFromCurrentSubreddit();

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            action: 'unsub',
            actionSource: '',
            srName: 'askReddit',
            sr: '',
            skipInitialDefaults: false,
          },
          context.metadata
        );
      });
    });
  });

  describe('Subreddit api model', () => {
    test('getCommentsAndPostsByIds()', async () => {
      const subreddit = new Subreddit({
        id: subredditId,
        displayName: 'askReddit',
        subredditType: 'public',
        createdUtc: Date.parse('2022-01-03'),
        lang: 'en',
        allowedMediaInComments: [],
        userFlairRichtext: [],
        submissionType: 'any',
      });

      const spyPlugin = redditApiPlugins.LinksAndComments.Info;
      spyPlugin.mockImplementationOnce(async () => mockListingWithPostsAndComments);

      await runWithTestContext(async () => {
        const listing = subreddit.getCommentsAndPostsByIds(['t3_abc123', 't1_xyz123']);
        const items = await listing.all();

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            subreddits: ['t5_t5_abc123'],
            thingIds: ['t3_abc123', 't1_xyz123'],
          },
          context.metadata
        );

        expect(items).toMatchSnapshot();
      });
    });

    test('getUserFlair()', async () => {
      const subredditName = 'askReddit';
      const subreddit = createTestSub({ name: subredditName });

      const mockUserFliar = {
        flairCssClass: undefined,
        flairText: 'Moderator',
        user: username,
      };

      const mockFlairListResponse = {
        users: [mockUserFliar],
      };

      const spyPlugin = redditApiPlugins.Flair.FlairList;
      spyPlugin.mockImplementationOnce(async () => mockFlairListResponse);

      await runWithTestContext(async () => {
        const response = await subreddit.getUserFlair();

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            subreddit: subredditName,
          },
          context.metadata
        );

        expect(response.users).toEqual([
          {
            flairCssClass: mockUserFliar.flairCssClass,
            user: mockUserFliar.user,
            flairText: mockUserFliar.flairText,
          },
        ]);
      });
    });

    test('getUserFlair({ usernames: ["user1, user2"] })', async () => {
      const subredditName = 'askReddit';
      const subreddit = createTestSub({ name: subredditName });

      const mockUserFliar1 = {
        flairCssClass: undefined,
        flairText: 'Moderator',
        user: 'user1',
      };

      const mockUserFliar2 = {
        flairCssClass: undefined,
        flairText: 'Moderator',
        user: 'user2',
      };

      const spyPlugin = redditApiPlugins.Flair.FlairList;
      spyPlugin.mockImplementation(async ({ name }) => ({
        users: [
          {
            flairCssClass: undefined,
            flairText: 'Moderator',
            user: name,
          },
        ],
      }));

      await runWithTestContext(async () => {
        const response = await subreddit.getUserFlair({ usernames: ['user1', 'user2'] });

        expect(spyPlugin).toHaveBeenNthCalledWith(
          1,
          {
            subreddit: subredditName,
            name: 'user1',
          },
          context.metadata
        );

        expect(spyPlugin).toHaveBeenNthCalledWith(
          2,
          {
            subreddit: subredditName,
            name: 'user2',
          },
          context.metadata
        );

        expect(response.users).toEqual([
          {
            flairCssClass: mockUserFliar1.flairCssClass,
            user: mockUserFliar1.user,
            flairText: mockUserFliar1.flairText,
          },
          {
            flairCssClass: mockUserFliar2.flairCssClass,
            user: mockUserFliar2.user,
            flairText: mockUserFliar2.flairText,
          },
        ]);
      });
    });

    test('getModqueue()', async () => {
      const subredditName = 'askReddit';
      const subreddit = createTestSub({ name: subredditName });

      const spyPlugin = redditApiPlugins.Moderation.AboutLocation;
      spyPlugin.mockImplementationOnce(async () => mockListingWithPostsAndComments);

      await runWithTestContext(async () => {
        const listing = subreddit.getModQueue();
        await listing.all();

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            subreddit: subredditName,
            location: AboutLocations.Modqueue,
            type: 'all',
            limit: 100,
            after: undefined,
            before: undefined,
            more: undefined,
            only: undefined,
          },
          context.metadata
        );
      });
    });

    test('getModqueue({ type: "post" })', async () => {
      const subredditName = 'askReddit';
      const subreddit = createTestSub({ name: subredditName });

      const spyPlugin = redditApiPlugins.Moderation.AboutLocation;
      spyPlugin.mockImplementationOnce(async () => mockListingWithPostsAndComments);

      await runWithTestContext(async () => {
        const listing = subreddit.getModQueue({ type: 'post' });
        await listing.all();

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            subreddit: subredditName,
            location: AboutLocations.Modqueue,
            type: 'post',
            only: 'links',
            limit: 100,
            after: undefined,
            before: undefined,
            more: undefined,
          },
          context.metadata
        );
      });
    });

    test('getModqueue({ type: "comment" })', async () => {
      const subredditName = 'askReddit';
      const subreddit = createTestSub({ name: subredditName });

      const spyPlugin = redditApiPlugins.Moderation.AboutLocation;
      spyPlugin.mockImplementationOnce(async () => mockListingWithPostsAndComments);

      await runWithTestContext(async () => {
        const listing = subreddit.getModQueue({ type: 'comment' });
        await listing.all();

        expect(spyPlugin).toHaveBeenCalledWith(
          {
            subreddit: subredditName,
            location: AboutLocations.Modqueue,
            type: 'comment',
            only: 'comments',
            limit: 100,
            after: undefined,
            before: undefined,
            more: undefined,
          },
          context.metadata
        );
      });
    });
  });
});
