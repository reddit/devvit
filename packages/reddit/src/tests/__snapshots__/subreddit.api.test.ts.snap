// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Subreddit API > RedditClient:Subreddit > getCommentsAndPostsByUser() 1`] = `
[
  {
    "approved": false,
    "authorName": "unusual_setup",
    "body": "This is a comment",
    "collapsedBecauseCrowdControl": false,
    "createdAt": +053971-01-26T00:00:00.000Z,
    "distinguishedBy": undefined,
    "edited": false,
    "id": "t1_abc123",
    "ignoringReports": false,
    "locked": false,
    "modReportReasons": [],
    "numReports": 0,
    "parentId": "t3_abc123",
    "permalink": "permalink://",
    "postId": "t3_abc123",
    "removed": false,
    "replies": Listing {
      "_fetch": [Function],
      "children": [],
      "hasMore": true,
      "limit": Infinity,
      "pageSize": 100,
    },
    "score": 0,
    "spam": false,
    "stickied": false,
    "subredditId": "t5_abc123",
    "subredditName": "pics",
    "url": "permalink://",
    "userReportReasons": [],
  },
  {
    "approved": false,
    "archived": false,
    "authorId": undefined,
    "authorName": "unusual_setup",
    "body": undefined,
    "bodyHtml": undefined,
    "createdAt": +053973-10-22T00:00:00.000Z,
    "distinguishedBy": undefined,
    "edited": false,
    "flair": {
      "backgroundColor": undefined,
      "cssClass": undefined,
      "richtext": [],
      "templateId": undefined,
      "text": undefined,
      "textColor": undefined,
      "type": undefined,
    },
    "hidden": false,
    "id": "t3_abc123",
    "ignoringReports": false,
    "locked": false,
    "modReportReasons": [],
    "nsfw": false,
    "numberOfComments": 0,
    "numberOfReports": 0,
    "permalink": "permalink://",
    "quarantined": false,
    "removed": true,
    "removedBy": "spez",
    "removedByCategory": "moderator",
    "score": 0,
    "secureMedia": undefined,
    "spam": false,
    "spoiler": false,
    "stickied": false,
    "subredditId": "t5_abc123",
    "subredditName": "pics",
    "thumbnail": undefined,
    "title": "This is a post",
    "url": "url://",
    "userReportReasons": [],
  },
]
`;

exports[`Subreddit API > RedditClient:Subreddit > getSubredditRemovalReasons() 1`] = `
[
  {
    "id": "uuid-abc",
    "message": "This is spam!",
    "title": "Spam",
  },
]
`;

exports[`Subreddit API > Subreddit api model > getCommentsAndPostsByIds() 1`] = `
[
  {
    "approved": false,
    "authorName": "unusual_setup",
    "body": "This is a comment",
    "collapsedBecauseCrowdControl": false,
    "createdAt": +053971-01-26T00:00:00.000Z,
    "distinguishedBy": undefined,
    "edited": false,
    "id": "t1_abc123",
    "ignoringReports": false,
    "locked": false,
    "modReportReasons": [],
    "numReports": 0,
    "parentId": "t3_abc123",
    "permalink": "permalink://",
    "postId": "t3_abc123",
    "removed": false,
    "replies": Listing {
      "_fetch": [Function],
      "children": [],
      "hasMore": true,
      "limit": Infinity,
      "pageSize": 100,
    },
    "score": 0,
    "spam": false,
    "stickied": false,
    "subredditId": "t5_abc123",
    "subredditName": "pics",
    "url": "permalink://",
    "userReportReasons": [],
  },
  {
    "approved": false,
    "archived": false,
    "authorId": undefined,
    "authorName": "unusual_setup",
    "body": undefined,
    "bodyHtml": undefined,
    "createdAt": +053973-10-22T00:00:00.000Z,
    "distinguishedBy": undefined,
    "edited": false,
    "flair": {
      "backgroundColor": undefined,
      "cssClass": undefined,
      "richtext": [],
      "templateId": undefined,
      "text": undefined,
      "textColor": undefined,
      "type": undefined,
    },
    "hidden": false,
    "id": "t3_abc123",
    "ignoringReports": false,
    "locked": false,
    "modReportReasons": [],
    "nsfw": false,
    "numberOfComments": 0,
    "numberOfReports": 0,
    "permalink": "permalink://",
    "quarantined": false,
    "removed": true,
    "removedBy": "spez",
    "removedByCategory": "moderator",
    "score": 0,
    "secureMedia": undefined,
    "spam": false,
    "spoiler": false,
    "stickied": false,
    "subredditId": "t5_abc123",
    "subredditName": "pics",
    "thumbnail": undefined,
    "title": "This is a post",
    "url": "url://",
    "userReportReasons": [],
  },
]
`;
